### DISCUSSION TITLE
App crash on setModel method of ModelNote class

### DISCUSSION THREAD

**Discussion by DerkachYuriy** [sceneview/sceneview#164] [Q&A]
Hello 
I don't really understand why my app crashes on setModel method:

```
    private fun addObjectToSceneProcessHit(hit: HitResult, sceneView: ArSceneView?, objectRenderable: Model): ArModelNode? {
        Log.i(TAG, "Process hit")
        val frame = sceneView?.currentFrame
        if (frame?.camera?.trackingState == TrackingState.TRACKING) {
            val anchor = hit.createAnchor()
            val anchorNode = ArModelNode(placementMode = PlacementMode.PLANE_HORIZONTAL)
            anchorNode.isSmoothPoseEnable = false
            anchorNode.anchor = anchor
            anchorNode.parent = sceneView
            anchorNode.setModel(objectRenderable, autoAnimate = false)
            return anchorNode
        }
        return null
    }
```
App crashes in this row:
`anchorNode.setModel(objectRenderable, autoAnimate = false)`
when this method tries to get model.entities inside ModelNote class

**Comment by ThomasGorisse** [comment]
Hi,
Can you provide the crash logs?

**Comment by DerkachYuriy** [comment]
`   #00 pc 00000000001b8958  /data/app/~~ZEh-T8Jg1Vidux3vqfPNdw==/my.app.ar-d0HM0ey55S1HfLz-3yhltg==/lib/arm64/libgltfio-jni.so (filament::gltfio::FilamentAsset::getEntityCount() const+0) (BuildId: c6c850b752925f24cdd0cecc117fd277bdf218c6)`

This is all that I have, in this case

**Reply by ThomasGorisse** [reply]
The issue might come from your model.
Try to use one coming from the samples and see if you have an issue with it.

**Comment by DerkachYuriy** [comment]
I successfully run your "model-viewer" sample on my android device. But I found an issue with it. 
I wrote this code in Activity.kt: 
```
        supportFragmentManager.beginTransaction().addToBackStack("").replace(R.id.containerFragment, SomeFragment()).commit()
        supportFragmentManager.beginTransaction().addToBackStack("").replace(R.id.containerFragment, MainFragment()).commit()
```
It's the only change that I've made, just to create another simple fragment with the text, to test how well MainFragment() will finish it's work by onBackPressed. 

And test wasn't successful - app got crash:  

```
18:44:25.943  A  Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x6d0 in tid 4843 (ple.modelviewer), pid 4843 (ple.modelviewer)
18:44:26.207  A  Cmdline: io.github.sceneview.sample.modelviewer
18:44:26.207  A  pid: 4843, tid: 4843, name: ple.modelviewer  >>> io.github.sceneview.sample.modelviewer <<<
18:44:26.207  A        #00 pc 00000000001b7e2c  /data/app/~~4MO7NR8l9jfzURjlQWMNQg==/io.github.sceneview.sample.modelviewer-T_JMG08aaRInIVgX8qtKOQ==/base.apk!libgltfio-jni.so (BuildId: c6c850b752925f24cdd0cecc117fd277bdf218c6)
```

Help me please with that. Because my app that uses your library has the same architecture - a lot of fragments in a single activity. And the fragment with your io.github.sceneview.SceneView is one of them.

**Comment by ThomasGorisse** [comment]
The `setModel` method is from older SceneView versions. In v4.0.0-rc.1, model loading is handled via `rememberModelInstance`:

```kotlin
val model = rememberModelInstance(modelLoader, "models/mymodel.glb")
if (model != null) {
    ModelNode(modelInstance = model)
}
```

This handles threading correctly (Filament JNI calls must run on the main thread) and returns `null` while loading. Common crash causes in older versions were calling model loading from background threads — `rememberModelInstance` handles this safely.
